synchronization with external user registry

FND opened the pull request:
this is required to display authors' full names - see commit messages for details

this is work in progress and not yet ready to be merged: 048e2532d4e55c4a13abad73f9c46c0634794f83 illustrates that we probably want to properly associate posts (`Entry` / `WeeklyStatus` / `Comment`) with the respective `User` (i.e. create model relationships)

any volunteers for that part? /cc @mjansing @mrreynolds @aknrdureegaesr

FND commented:
associations created in 386aed1d76a6de890f481124a48d785226725b13 (note the commit message) - should be ready to merge now, though I'd appreciate a review (also WRT eager loading)

aknrdureegaesr commented:
At first glance, I'm missing the documentation how to exercise this.

It would be good to document the environmental variables `IBLOG_USERLIST_*` and to describe the format of the JSON file expected at that URL.

I'd also like to see a sample JSON file checked in here, and the ability to use such a file for testing purposes.

(If you're not in the mood to do this, I'd offer to provide a fix that addresses these issues myself.)

aknrdureegaesr commented on app/models/user.rb:
```diff
@@ -0,0 +1,20 @@
+# encoding: UTF-8
+# Copyright 2014 innoQ Deutschland GmbH
```
It has been 2015 for a while now.

FND replied:
I'd intentionally copied the existing header, leaving a proper update of all license headers for a separate commit.

aknrdureegaesr replied:
Fine with me. Do you intend that commit be part of this pull request before merging?

FND replied:
I'd conveniently left out that I wasn't gonna bother myself, as I don't much care for copyright... Even if I did, updating the license headers every year seems unnecessary - once per decade would probably suffice.

FND commented:
> I'd offer to provide a fix that addresses these issues myself

That would be most welcome!

The expected JSON format looks like this:

```json
{
    "members": {
        "jd": {
            "displayName": "John Doe",
            "avatar": "/avatars/jd",
            "avatar_small": "/avatars/jd/32x32",
        }, {
            ...
        }
    }
}
```

(avatars are not currently being used, but these should eventually replace the [hard-coded URI](https://github.com/innoq/iBlog/blob/64e81ee64783b88dcf67dcc9445d1db8187873c7/app/helpers/comments_helper.rb#L16))

I suppose the reason I didn't document any of this myself is broken windows; it feels like this application is held together by spit and bubble gum (see also [notifier](https://github.com/innoq/iBlog/blob/64e81ee64783b88dcf67dcc9445d1db8187873c7/app/services/notifier.rb)), though I suppose part of that might be due to OCD and my general dislike of Rails.

aknrdureegaesr commented:
I'd like to see the userdata update functionality available as a `rake` task.

aknrdureegaesr commented:
Probably rename the new table from `user` to `author`?

aknrdureegaesr commented:
This is unusual. We could make it more usual as follows:

* Throw out the `author` column from tables like `entries` or `comments`.
* Replace that with a new foreign key `author_id` to the `author` table.
* When an entry, comment or whatever comes in and we do not have an author corresponding to the handle in the author table, create a new line in the `author` table (initially setting `name` to the same value as `handle`).
* Write a smart migration that migrates the legacy data.

The idea here is, let's cover up the fact that originally, iBlog did not have authors.  This shines through clearly through the pull request as we have it now. Improve on it so it looks like authors have always been entities of their own in iBlog. 

@FND - what do you think?

FND commented:
> Probably rename the new table from `user` to `author`?

The intent there was to provide a user registry without any presuppositions about how they're being used. Basically a loosely coupled service, an optional one at that.

I'm quite aware that this approach of augmenting a model's attribute (post author) via another model (user) is unusual. At the time, it seemed like the least invasive way to implement this feature and saved me from having to restructure model relationships (which seemed like a lot of tedious work in various places) - i.e. it was a conscious compromise to minimize effort. You're very welcome to change that though if you're up for it.

(Note that I doubt anyone outside innoQ is using iBlog, so the user registry doesn't actually need to be optional.)

aknrdureegaesr commented:
I'm already working at moving "authors" to become first class citizens in our object model, in the add_author branch. Some of the tedious work has already been done.

aknrdureegaesr commented:
superseeded by #205
